Extend subgraph utils to cover merging a node following a subgraph

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* **#52513 Extend subgraph utils to cover merging a node following a subgraph**
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first.

Differential Revision: [D26696697](https://our.internmc.facebook.com/intern/diff/D26696697)

Commit:
Extend subgraph utils to cover merging a node following a subgraph

[ghstack-poisoned]

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -26,9 +26,11 @@ std::vector<c10::optional<const Use>> gatherLastUses(
 // outside of the subgraph to be merged into we do not need to track.
 struct ValueMapper {
-  ValueMapper(Node* to_merge, AliasDb& db, size_t subgraph_num_outputs) {
+  ValueMapper(Node* to_merge, AliasDb& db, c10::optional<Node*> existing) {
```
could we relate the comment below to `ValueMapper` c-tor's parameters?

eellison replied:
i'm not sure what you mean

ZolotukhinM replied:
I'm not sure what @Krovatkin meant, but I'd find it helpful if we had a description of what each argument of the constructor is used for (especially the last one).

Krovatkin commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -96,28 +103,29 @@ Node* executeSubgraphMergeAndUpdateAliasing(
 
-  // defer destroying nodes until after all nodes have been merged,
-  // to make iterators easier to reason about
   std::vector<Node*> merged_nodes;
```
looks like this isn't used anymore?

Krovatkin commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -96,28 +103,29 @@ Node* executeSubgraphMergeAndUpdateAliasing(
-
-  for (Node* n : merged_nodes) {
-    n->destroy();
+    mergeNodeIntoSubgraph(node, mergeTo);
```
` /*destroyNode*/ false` do you know why this was an issue before?

eellison replied:
this could have been removed with the removal of the vmap API PR, I think it existed for that reason

Krovatkin commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -233,68 +253,77 @@ void mergeNodeIntoSubgraph(
-      groupOutput->copyMetadata(oldOutput);
-      oldOutput->replaceAllUsesWith(groupOutput);
-    }
+    auto newOutput = mergedNode->outputs()[i];
```
why did we hoist pruning unused outputs outside and after this loop?

eellison replied:
Because we want to do it after we destroy the node

ZolotukhinM replied:
Might be worth adding a comment saying exactly that :)

Krovatkin commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -233,68 +253,77 @@ void mergeNodeIntoSubgraph(
+    auto inputs = subgraphNode->inputs();
+    for (size_t i = 0; i < toMerge->outputs().size(); ++i) {
+      auto it = std::find(inputs.begin(), inputs.end(), toMerge->outputs()[i]);
+      if (it != inputs.end()) {
```
I understand you didn't change this code but do you know when we can have `it == inputs.end()`?

eellison replied:
> If n's outputs were inputs to `group`, remove them since we just merged 

If any of `n`'s outputs weren't inputs to the subgraph.

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Krovatkin reviewed (commented):
a few questions

ZolotukhinM commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -202,20 +212,30 @@ void mergeNodeIntoSubgraph(
 
-  // Map from values in the surrounding graph to inputs in the subgraph
-  std::unordered_map<Value*, Value*> inputsMap;
+  // Map from values in the surrounding graph to values in the subgraph
```
Minor: tbh, this comment is hard to follow. Maybe add an example or change "values" to "inputs/outputs".

ZolotukhinM reviewed (approved):
Looks good to me, but please make sure @Krovatkin doesn't have any concerns regarding that either!

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first. 

[ghstack-poisoned]

Commit:
Update on "Extend subgraph utils to cover merging a node following a subgraph"


Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first.

Differential Revision: [D26696697](https://our.internmc.facebook.com/intern/diff/D26696697)

[ghstack-poisoned]
